qiskit/aqua/components/neural_networks/quantum_generator.py

```diff
@@ -308,6 +310,45 @@ def objective_function(params):
 
         return objective_function
 
+    def _get_analytical_loss_gradient(self, quantum_instance, discriminator):
+        """
+        Get analytical loss gradient
+
+        Args:
+            quantum_instance (QuantumInstance): used to run the quantum circuit.
+            discriminator (torch.nn.Module): discriminator network to compute the sample labels.
+
+        Returns:
+            analytical_loss_gradient: function that computes analytical loss gradients.
+        """
+
+        def analytical_loss_gradient(initial_point):
+            """
+            Analytical loss gradient
+
+            Args:
+                initial_point (np.ndarray): Initial values for the variational parameters.
+
+            Returns:
+                loss_gradients (np.ndarray): array of partial derivatives of the loss
+                    function w.r.t. the variational parameters.
+            """
+            # bound_params = self._bound_parameters
+            free_params = self._free_parameters
+            generated_data, _ = self.get_output(quantum_instance,
+                                                             params=initial_point,
+                                                             shots=self._shots)
+            prediction_generated = discriminator.get_label(generated_data, detach=True)
+            op = ~CircuitStateFn(primitive=self.generator_circuit)
+            grad_object = Gradient(grad_method='param_shift').convert(operator=op,
```

Comment:
It would be ideal if the user can choose the method for computing the gradient and if none is given the standard use should be finite difference. As @woodsp-ibm mentioned the [VQE algorithm](https://github.com/Qiskit/qiskit-aqua/blob/master/qiskit/aqua/algorithms/minimum_eigen_solvers/vqe.py) offers a good example.

Reply:
@Zoufalc got it, thanks - I'll rework the QGAN algorithm so that we can pass in a GradientBase object / function that returns gradients which will be used to compute gradients analytically. Here's what I understood, please let me know if this is right:
* If an optimizer and a GradientBase object / gradient function is passed in, use both of them.
* If only as optimizer is passed in, but no GradientBase object / gradient function, use numeric gradients
* If only a GradientBase object / gradient function is provided (is this a case we should account for?), use ADAM + analytic gradients
* If neither are given, use ADAM + numeric gradients.

Another thing I noticed was that the VQE algorithm lets you change optimizers / gradient functions when you call [find_minimum](https://qiskit.org/documentation/stubs/qiskit.aqua.algorithms.VQE.html#qiskit.aqua.algorithms.VQE.find_minimum) - I think this would be a cool addition to the QGAN algorithm (I raised a new issue for this [here](#1454))

Reply:
Yes, that's right.

This would be indeed a nice extension. Thanks.